File: src/Controller/Controller.java
package Controller;

import DataTypes.*;
import Model.ProgramState;
import Model.Statement.IStatement;
import Repository.IRepository;
import Exception.MyStatementExecutionException;
import Exception.EmptyMyStackException;

import java.util.Map;
import java.util.Collection;
import java.util.stream.Collectors;

/**
 * Created by vladc on 25.10.2016.
 */
public class Controller {
    private IRepository repo;

    public Controller(IRepository repo) {
        this.repo = repo;
    }

    public void add(IStatement st) {
        ProgramState state = new ProgramState(
                new MyStack<>(),
                new MyDictionary<>(),
                new MyList<>(),
                new MyDictionary<>(),
                new MyHeap(),
                st
        );
        this.repo.add(state);
    }

    public ProgramState getCurrentProgram() {
        return repo.getCurrentProgram();
    }

    public ProgramState oneStep(ProgramState state) throws MyStatementExecutionException {
        MyIStack<IStatement> st = state.getExeStack();

        try {
            IStatement currentStatement = st.pop();
            return currentStatement.execute(state);
        }
        catch (EmptyMyStackException e) {
            throw new MyStatementExecutionException("Exception. Execution Stack is empty.");
        }
    }

    public void allStep() throws MyStatementExecutionException {
        ProgramState program = repo.getCurrentProgram();

        while (!program.getExeStack().isEmpty()) {
            oneStep(program);
            program.getHeap().setContent(conservativeGarbageCollector(
                    program.getSymTable().values(),
                    program.getHeap().getContent()
            ));
            System.out.println(program.toString());
            repo.logProgramStateExec();
        }
        repo.logEndOfExecution();
    }

    private Map<Integer, Integer> conservativeGarbageCollector(
            Collection<Integer> symTableValues,
            Map<Integer, Integer> heap
    ) {
        return heap.entrySet().stream()
                .filter(e -> symTableValues.contains(e.getKey()))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }
}

File: Toy-Language-Interpreter/Controller/Controller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Toy_Language_Interpreter.Repository;
using Toy_Language_Interpreter.Model;
using Toy_Language_Interpreter.Model.Statement;
using Toy_Language_Interpreter.DataTypes;
using Toy_Language_Interpreter.MyException;

using System.IO;

namespace Toy_Language_Interpreter.Controller
{
    class MyController
    {
        private IRepository repo;

        public MyController(IRepository repo)
        {
            this.repo = repo;
        }

        public void Add(IStatement st)
        {
            ProgramState state = new ProgramState(new MyStack<IStatement>(), new MyDictionary<string, int>(),
                new MyList<int>(), new MyDictionary<int, MyPair<string, StreamReader>>(), st);
            this.repo.Add(state);
        }

        public ProgramState GetCurrentProgram()
        {
            return this.repo.GetCurrentProgram();
        }

        public ProgramState OneStep(ProgramState state)
        {
            MyIStack<IStatement> st = state.GetExeStack();

            try
            {
                IStatement currentStatement = st.Pop();
                return currentStatement.Execute(state);
            }
            catch (EmptyMyStackException)
            {
                throw new MyStatementExecutionException("Exception. Execution stack is empty.");
            }
        }

        public void AllStep()
        {
            ProgramState program = repo.GetCurrentProgram();

            while (!program.GetExeStack().IsEmpty())
            {
                OneStep(program);
                repo.LogProgramStateExec();
            }
        }
    }
}
